pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,244 @@
+func (r *UnsandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db
+	if r.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly || !r.Vdb.IsEON() {
+		return ctrl.Result{}, nil
+	}
+
+	// collect pod facts for main cluster
+	if err := r.PFacts.Collect(ctx, r.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// reconcile sandbox status for the subclusters that are already unsandboxed
+	if err := r.reconcileSandboxStatus(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// reconcile the sandbox config map if it expires
+	err, deleted := r.reconcileSandboxConfigMap(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// if we deleted the expired config map, we don't need to make unsandbox operation any more
+	if deleted {
+		return ctrl.Result{}, nil
+	}
+
+	// only execute unsandbox op when unsandbox trigger id and sandbox name are set
+	if r.ConfigMap.Annotations[vmeta.SandboxControllerUnsandboxTriggerID] != "" &&
+		r.ConfigMap.Data[vapi.SandboxNameKey] != "" {
+		// we ignore the config map that contains wrong data,
+		// this kind of config map could be old or created by others
+		if r.ConfigMap.Data[vapi.VerticaDBNameKey] != r.Vdb.Name {
+			r.Log.Info("Vdb name in sandbox config map doesn't match current vdb's, skip unsandbox operation for this config map",
+				"configMapName", r.ConfigMap.Name, "vdbNameInConfigMap", r.ConfigMap.Data[vapi.VerticaDBNameKey], "currentVdbName", r.Vdb.Name)
+			return ctrl.Result{}, nil
+		}
+		return r.unsandboxSubclusters(ctx)
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileSandboxStatus will update sandbox status for the subclusters that are already unsandboxed
+func (r *UnsandboxSubclusterReconciler) reconcileSandboxStatus(ctx context.Context) error {
+	scSbInStatus := make(map[string]string)
+	for _, sb := range r.Vdb.Status.Sandboxes {
+		for _, sc := range sb.Subclusters {
+			scSbInStatus[sc] = sb.Name
+		}
+	}
+	sbScMap := r.PFacts.FindUnsandboxedSubclustersStillInSandboxStatus(scSbInStatus)
+	for sb, scs := range sbScMap {
+		err := r.updateSandboxStatus(ctx, sb, scs)
+		if err != nil {
+			r.Log.Error(err, "failed to update sandbox status", "sandbox", sb, "new subclusters", scs)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// reconcileSandboxConfigMap will delete sandbox config map if it expires
+func (r *UnsandboxSubclusterReconciler) reconcileSandboxConfigMap(ctx context.Context) (error, bool) {
+	foundSbInStatus := false
+	for _, sb := range r.Vdb.Status.Sandboxes {
+		if sb.Name == r.ConfigMap.Data[vapi.SandboxNameKey] {
+			foundSbInStatus = true
+			break
+		}
+	}
+	// if we cannot find the sandbox in status, it means the sandbox has been unsandboxed
+	// or does not exist
+	if !foundSbInStatus {
+		cmName := r.ConfigMap.Name
+		err := r.Client.Delete(ctx, r.ConfigMap)
```

Comment:
I think unsandbox controller deleting the map will be faster and intuitive. After unsandbox controller unsandbox the subclusters, it knows if it can delete the config map. We don't need to wait for a loop in vdb controller. Before vdb controller runs into unsandbox_reconciler, the config map could be picked up again in sandbox controller, that will have some cost.

Reply:
I like that we delete the ConfigMap in here too. It's a clear distinction that once the ConfigMap is created, the sandbox controller is reponsible for its deletion. I noticed in the vdb controller we try to delete configMap's too. Is is necessary to delete it from there? Should we just have a single spot that does the deletion. That will be easier to reason about.

Reply:
Sure, I will remove the delete-config-map code in vdb controller.